### PR TITLE
#11 add jest

### PR DESCRIPTION
**Pull request by olya646**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,8 +4,8 @@
   "description": "",
   "main": "./build/index",
   "scripts": {
-    "test": "NODE_ENV=test mocha --compilers js:babel-core/register --require babel-polyfill --recursive test/.setup.js test/**/*.test.js",
-    "test-coverage": "NODE_ENV=test babel-node ./node_modules/.bin/babel-istanbul cover --include-all-sources --report html --dir coverage/html ./node_modules/.bin/_mocha -- --recursive test/.setup.js test/**/*.test.js",
+    "test": "NODE_ENV=test jest --compilers js:babel-core/register --require babel-polyfill --recursive test/.setup.js test/**/*.test.js",
```
Comment (ferrannp) [inline package.json]: You need to get rid of all this old configuration, that was just required by mocha. You just need:

```
"test": "jest"
```

File: package.json [hunk 2]
```diff
@@ -4,8 +4,8 @@
...
+    "test-coverage": "NODE_ENV=test babel-node ./node_modules/.bin/babel-istanbul cover --include-all-sources --report html --dir coverage/html ./node_modules/.bin/_jest -- --recursive test/.setup.js test/**/*.test.js",
```
Comment (ferrannp) [inline package.json]: All this `babel-istanbul` stuff should be also removed. Your are collecting now coverage with Jest. The command you need to run here is:

```
"test-coverage": "jest --coverage"
```

File: package.json [hunk 3]
```diff
@@ -28,21 +28,24 @@
...
     "babel-istanbul": "^0.11.0",
+    "babel-jest": "^18.0.0",
     "babel-polyfill": "6.16.0",
     "babel-preset-es2015": "6.18.0",
     "babel-preset-react": "6.16.0",
     "babel-preset-stage-0": "6.16.0",
     "babel-register": "6.18.0",
     "babel-runtime": "6.18.0",
     "chai": "3.5.0",
+    "chai-jest-snapshot": "^0.2.1",
```
Comment (ferrannp) [inline package.json]: We do not need this. We want to get rid of extra dependencies, you do not need `chai-jest-snapshot` neither `chai`.

File: package.json [hunk 4]
```diff
@@ -28,21 +28,24 @@
     "babel-cli": "^6.18.0",
     "babel-eslint": "7.1.0",
     "babel-istanbul": "^0.11.0",
```
Comment (ferrannp) [inline package.json]: We do not need `babel-istanbul` anymore now that Jest can collect coverage.

File: package.json [hunk 5]
```diff
@@ -28,21 +28,24 @@
...
     "babel-runtime": "6.18.0",
     "chai": "3.5.0",
+    "chai-jest-snapshot": "^0.2.1",
     "enzyme": "2.5.1",
     "eslint": "3.8.1",
     "eslint-plugin-react": "6.4.1",
     "flow-bin": "^0.33.0",
+    "jest": "^18.1.0",
     "jsdom": "^9.8.3",
     "mocha": "3.1.2",
```
Comment (ferrannp) [inline package.json]: We do not need `mocha`.

File: package.json [hunk 6]
```diff
@@ -28,21 +28,24 @@
...
     "babel-register": "6.18.0",
     "babel-runtime": "6.18.0",
     "chai": "3.5.0",
+    "chai-jest-snapshot": "^0.2.1",
     "enzyme": "2.5.1",
     "eslint": "3.8.1",
     "eslint-plugin-react": "6.4.1",
     "flow-bin": "^0.33.0",
+    "jest": "^18.1.0",
     "jsdom": "^9.8.3",
```
Comment (ferrannp) [inline package.json]: We do not need `jsdom`.

File: package.json [hunk 7]
```diff
@@ -55,5 +58,13 @@
...
     "react": ">=15.0.0",
     "react-dom": ">=15.0.0"
+  },
+  "jest": {
+      "unmockedModulePathPatterns": [
+      "node_modules/react/",
+      "node_modules/enzyme/"
+    ],
+    "collectCoverage": true,
+    "coverageFormats": ["json", "html"]
```
Comment (ferrannp) [inline package.json]: Probably you should remove `collectCoverage` and `coverageFormats`. For what I see, we just want to collect coverage when we run `npm run test-coverage`, not always.

File: test/components/FeatureFlagRenderer.test.js [hunk 8]
```diff
@@ -1,7 +1,6 @@
 import React from "react";
 import { shallow, mount } from "enzyme";
-import { expect } from "chai";
-import { spy, stub, sandbox } from "sinon";
+import sinon, { spy, stub, sandbox } from "sinon";
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: We need to get rid of sinon. To mock stuff, you can use [Jest mock api](https://facebook.github.io/jest/docs/mock-function-api.html#content).

File: test/components/FeatureFlagRenderer.test.js [hunk 9]
```diff
@@ -24,7 +23,7 @@ describe("components/FeatureFlagRenderer", () => {
         renderFeatureCallback={renderFeatureCallback}
       />
     );
-    expect(wrapper).to.exist;
+    expect(wrapper).toContainEqual;
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: `toContainEqual` expects a parameter, see http://facebook.github.io/jest/docs/expect.html#tocontainequalitem

> Use .toContainEqual when you want to check that an item is in a list. 

You do not want this here, you want:

```
expect(wrapper).toBeDefined();
```

File: test/components/FeatureFlagRenderer.test.js [hunk 10]
```diff
@@ -35,7 +34,7 @@ describe("components/FeatureFlagRenderer", () => {
         renderFeatureCallback={renderFeatureCallback}
       />
     );
-    expect(wrapper.find("[data-qa='FeatureFlag-my-test']")).to.exist;
+    expect(wrapper.find("[data-qa='FeatureFlag-my-test']")).toContainEqual;
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: All those `toContainEqual` are wrong, check http://facebook.github.io/jest/docs/expect.html.

File: test/components/FeatureFlagRenderer.test.js [hunk 11]
```diff
@@ -47,11 +46,11 @@ describe("components/FeatureFlagRenderer", () => {
         renderFeatureCallback={renderFeatureCallback}
       />
     );
-    expect(FeatureFlagRenderer.prototype.componentDidMount).to.have.property("callCount", 1);
+    expect(FeatureFlagRenderer.prototype.componentDidMount).toContainEqual;
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: `toContainEqual` is wrongly used.

File: test/lib/launchDarkly.test.js [hunk 12]
```diff
@@ -1,5 +1,5 @@
-import { expect } from "chai";
 import { spy } from "sinon";
+import { expect } from "chai";
```
Comment (ferrannp) [inline test/lib/launchDarkly.test.js]: We need to get rid of this so `jest` uses its on `expect`.

File: test/components/FeatureFlagRenderer.test.js [hunk 13]
```diff
@@ -169,25 +169,29 @@ describe("components/FeatureFlagRenderer", () => {
       stub(launchDarkly, "ldBrowserInit", ldClientStub);
     });
 
-    context("when the feature flag comes back true", () => {
+    afterEach(() => {
+      sinon.restore(launchDarkly);
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: We need to remove `sinon`.

File: test/components/FeatureFlagRenderer.test.js [hunk 14]
```diff
@@ -169,25 +169,29 @@ describe("components/FeatureFlagRenderer", () => {
...
-        expect(wrapper.state()).to.deep.equal({ checkFeatureFlagComplete: true, showFeature: false });
+        expect(wrapper.state()).toEqual({ checkFeatureFlagComplete: true, showFeature: false });
       });
     });
 
     describe("query param flag overrides if not undefined", () => {
       let _sandbox;
-      before(() => {
+      beforeEach(() => {
         _sandbox = sandbox.create();
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: We need to get rid of this in favour to Jest mocking system.

File: test/components/FeatureFlagRenderer.test.js [hunk 15]
```diff
@@ -197,29 +201,28 @@ describe("components/FeatureFlagRenderer", () => {
         variation.returns(true);
         sandbox.stub(launchDarkly, "getLocation").returns("http://ab.cdef.com?features.my-test=false");
```
Comment (ferrannp) [inline test/components/FeatureFlagRenderer.test.js]: With Jest you will do:

```
launchDarkly.getLocation = jest.fn().mocakImplementation(() => "http://ab.cdef.com?features.my-test=false");
```

**Comment by olya646** [comment]
Its working for me, all the tests are passed.
http://prntscr.com/e2kvj0

File: test/components/FeatureFlagRenderer.test.js (hunk 8 above)
Comment (olya646) [inline test/components/FeatureFlagRenderer.test.js]: We dont need sinon anymore.

File: test/components/LaunchDarkly.test.js [hunk 16]
```diff
@@ -17,18 +15,18 @@ describe("components/LaunchDarkly", () => {
     );
 
     const broadcast = subject.find(Broadcast);
-    expect(broadcast.prop("channel")).to.equal(BROADCAST_CHANNEL);
+    expect(broadcast.prop("channel")).toEqual(BROADCAST_CHANNEL);
   });
 
-  it("should pass the clientId and user as the value to the broadcast", () => {
+  fit("should pass the clientId and user as the value to the broadcast", () => {
```
Comment (ferrannp) [inline test/components/LaunchDarkly.test.js]: Remember to remove `fit` when using them. If you ran the tests, you will see 3 skipped tests with this.

**Comment by ferrannp** [comment]
@olya646 feel free to send a PR to TrueCar.

**Pull request closed without merging**